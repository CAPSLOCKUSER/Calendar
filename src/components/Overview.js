import React from 'react';
import {TIME_RANGE_MIN, TIME_RANGE_MAX} from '../constants';
import dateFormat from 'dateformat';
import leftPad from 'left-pad'; // :D
import Appointment from '../components/Appointment';
import {timeNodeHeight} from '../../style/values.scss';

const ONE_HOUR_HEIGHT = parseInt(timeNodeHeight, 10);

const timeListLength = TIME_RANGE_MAX - TIME_RANGE_MIN + 1;
const timeList = Array.from({ length: timeListLength }).map((_, index) => {
  const hour = TIME_RANGE_MIN + index;
  return `${leftPad(hour, 2, 0)}:00`;
});

function groupByOverlap() {
  let currentEndTime = null;
  return (prev, curr) => {
    if (currentEndTime === null) {
      currentEndTime = curr.endTime;
      return [[curr]];
    }

    if (currentEndTime > curr.startTime) {
      prev[prev.length - 1].push(curr);
    } else {
      prev.push([curr]);
    }

    if (currentEndTime < curr.endTime) {
      currentEndTime = curr.endTime;
    }

    return prev;
  }
}

function calculateGaps() {
  let previousEndTime = TIME_RANGE_MIN * 60;
  return group => {
    const result = group.map(data => {
      const height = Math.max(30, (data.endTime - data.startTime) / 60 * ONE_HOUR_HEIGHT);
      return {
        data,
        marginTop: (data.startTime - previousEndTime) / 60 * ONE_HOUR_HEIGHT,
        height,
      };
    });

    previousEndTime = Math.max(...result.map(({ data : { endTime } }) => endTime));

    return result;
  }
}

function stableSortByStart(a, b) {
  if (a.startTime === b.startTime) {
    return a.id - b.id;
  }

  return a.startTime - b.startTime;
}

const Overview = ({ appointments, onSelect, selectedId }) => {
  const now = new Date();
  const today = dateFormat(now, 'd mmmm yyyy').toLowerCase();

  const appointmentsGroupedByOverlap = appointments
    .sort(stableSortByStart)
    .reduce(groupByOverlap(), [])
    .map(calculateGaps());

  let container;
  const selectFirstAppointment = () => {
    const first = container.querySelector('.appointment');
    if (first) {
      first.focus();
    }
  };

  return (
    <div className="overview" onClick={() => onSelect(null)}>
      <header>{today}</header>
      <article ref={x => container = x}>
        <ul>
          {timeList.map((text, index) => <li key={index}>{text}</li>)}
        </ul>
        <div className="appointment-container">
          {appointmentsGroupedByOverlap.map((appointments, index) => (
            <div key={index} className="appointment-group">
              {appointments.map(({ data, marginTop, height }, index) => (
                <Appointment
                  key={index}
                  values={data}
                  style={{ marginTop, height }}
                  onSelect={onSelect}
                  isSelected={selectedId === data.id}
                />
              ))}
            </div>
          ))}
        </div>
      </article>
      <a
        href="#"
        onFocus={selectFirstAppointment}
        onClick={event => event.preventDefault()}
      />
    </div>
  );
};

export default Overview;
