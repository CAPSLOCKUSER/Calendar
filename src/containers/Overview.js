import React from 'react';
import { connect } from 'react-redux';
import { TIME_RANGE_MIN, TIME_RANGE_MAX } from '../constants';
import dateFormat from 'dateformat';
import leftPad from 'left-pad'; // :D
import Appointment from '../components/Appointment';
import { timeNodeHeight } from '../../style/values.scss';

const ONE_HOUR_HEIGHT = parseInt(timeNodeHeight, 10);

const timeListLength = TIME_RANGE_MAX - TIME_RANGE_MIN + 1;
const timeList = Array.from({ length: timeListLength }).map((_, index) => {
  const hour = TIME_RANGE_MIN + index;
  return `${leftPad(hour, 2, 0)}:00`;
});

function createGroupper() {
  let currentEndTime = null;
  return (prev, curr) => {
    if (currentEndTime === null) {
      currentEndTime = curr.endTime;
      return [[curr]];
    }

    if (currentEndTime > curr.startTime) {
      prev[prev.length - 1].push(curr);
    } else {
      prev.push([curr]);
    }

    if (currentEndTime < curr.endTime) {
      currentEndTime = curr.endTime;
    }

    return prev;
  }
}

function calculateGaps() {
  let previousEndTime = null;
  return group => {
    if (previousEndTime === null) {
      previousEndTime = TIME_RANGE_MIN * 60;
    }
    const groupStartTime = group[0].startTime;
    const topGapTime = groupStartTime - previousEndTime;
    const marginTop = topGapTime / 60 * ONE_HOUR_HEIGHT;
    previousEndTime = Math.max(...group.map(({ endTime }) => endTime));
    return {
      marginTop,
      appointments: group.map(data => {
        return Object.assign({}, data, {
          marginTop: (data.startTime - groupStartTime) / 60 * ONE_HOUR_HEIGHT,
          height: (data.endTime - data.startTime) / 60 * ONE_HOUR_HEIGHT,
        });
      }),
    }
  }
}

let Overview = ({ appointments }) => {
  const now = new Date();
  const today = dateFormat(now, 'd mmmm yyyy').toLowerCase();

  const appointmentsGroupedByOverlap = appointments
    .sort(({ startTime: a }, { startTime: b }) => a - b)
    .reduce(createGroupper(), [])
    .map(calculateGaps());

  console.log(appointmentsGroupedByOverlap);

  return (
    <div className="overview">
      <header>{today}</header>
      <article>
        <ul>
          {timeList.map((text, index) => <li key={index}>{text}</li>)}
        </ul>
        <div className="appointment-container">
          {appointmentsGroupedByOverlap.map(({ appointments, marginTop }, index) => (
            <div key={index} className="appointment-group" style={{ marginTop }}>
              {appointments.map((data, index) => (
                <Appointment key={index} {...data} />
              ))}
            </div>
          ))}
        </div>
      </article>
    </div>
  );
};

const mapStateToProps = ({ appointments }) => {
  return {
    appointments,
  }
};

const mapDispatchToProps = (dispatch) => {
  return {
    onTodoClick: (id) => {
      //dispatch(toggleTodo(id))
    }
  }
};

Overview = connect(
  mapStateToProps,
  mapDispatchToProps
)(Overview);

export default Overview;
