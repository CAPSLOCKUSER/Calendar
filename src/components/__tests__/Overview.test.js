import React from 'react';
import renderer from 'react-test-renderer';
import MockDate from 'mockdate';
import Overview from '../Overview';

jest.mock('../../../style/values.scss', () => ({
  timeNodeHeight: '30px',
}));

describe('components/Overview', () => {
  beforeEach(() => {
    MockDate.set('8/9/2016');
  });

  afterEach(() => {
    MockDate.reset();
  });

  it('renders correctly', () => {
    const exampleAppointments = require('./example_appointments.json');
    const tree = renderer.create(
      <Overview
        appointments={exampleAppointments}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
