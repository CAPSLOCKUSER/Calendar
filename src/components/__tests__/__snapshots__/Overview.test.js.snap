exports[`components/Overview renders correctly 1`] = `
<div
  className="overview"
  onClick={[Function onClick]}>
  <header>
    9 august 2016
  </header>
  <article>
    <ul>
      <li>
        07:00
      </li>
      <li>
        08:00
      </li>
      <li>
        09:00
      </li>
      <li>
        10:00
      </li>
      <li>
        11:00
      </li>
      <li>
        12:00
      </li>
      <li>
        13:00
      </li>
      <li>
        14:00
      </li>
      <li>
        15:00
      </li>
      <li>
        16:00
      </li>
      <li>
        17:00
      </li>
    </ul>
    <div
      className="appointment-container">
      <div
        className="appointment-group"
        style={
          Object {
            "marginTop": 0
          }
        }>
        <a
          className="appointment"
          href="#"
          onClick={[Function onClick]}
          onFocus={[Function onFocus]}
          style={
            Object {
              "height": 15,
              "marginTop": 0
            }
          }>
          <h1>
            iuf
          </h1>
          <div
            className="time">
            07:00 - 07:30
          </div>
          <br />
          <p
            className="description">
            sdf
          </p>
        </a>
      </div>
      <div
        className="appointment-group"
        style={
          Object {
            "marginTop": 15
          }
        }>
        <a
          className="appointment"
          href="#"
          onClick={[Function onClick]}
          onFocus={[Function onFocus]}
          style={
            Object {
              "height": 225,
              "marginTop": 0
            }
          }>
          <h1>
            sdfffff
          </h1>
          <div
            className="time">
            08:00 - 15:30
          </div>
          <br />
          <p
            className="description">
            sdf
          </p>
        </a>
        <a
          className="appointment"
          href="#"
          onClick={[Function onClick]}
          onFocus={[Function onFocus]}
          style={
            Object {
              "height": 105,
              "marginTop": 0
            }
          }>
          <h1>
            sfg
          </h1>
          <div
            className="time">
            08:00 - 11:30
          </div>
          <br />
          <p
            className="description">
            asdf
          </p>
        </a>
        <a
          className="appointment"
          href="#"
          onClick={[Function onClick]}
          onFocus={[Function onFocus]}
          style={
            Object {
              "height": 128,
              "marginTop": 157
            }
          }>
          <h1>
            sdff
          </h1>
          <div
            className="time">
            13:14 - 17:30
          </div>
          <br />
          <p
            className="description">
            sdsd
          </p>
        </a>
      </div>
    </div>
  </article>
  <a
    href="#"
    onClick={[Function onClick]}
    onFocus={[Function selectFirstAppointment]} />
</div>
`;
